File: src/main/java/net/katsstuff/danmakucore/network/SpellcardInfoPacket.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network;

import java.util.UUID;

import io.netty.buffer.ByteBuf;
import net.katsstuff.danmakucore.DanmakuCore;
import net.katsstuff.danmakucore.entity.spellcard.spellcardbar.SpellcardInfo;
import net.minecraft.client.Minecraft;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.fml.common.network.ByteBufUtils;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class SpellcardInfoPacket {

	public enum Action {
		ADD,
		SET_NAME,
		SET_MIRROR,
		SET_POS,
		SET_POS_ABSOLUTE,
		REMOVE
	}

	public static class Message implements IMessage {

		private UUID uuid;
		private ITextComponent name;
		private Action action;

		private boolean mirror;

		private float posX;
		private float posY;

		private float prevPosX;
		private float prevPosY;


		public Message(SpellcardInfo info, Action action) {
			this.uuid = info.getUuid();
			this.name = info.getName();
			this.action = action;

			mirror = info.shouldMirrorText();

			posX = info.getPosX();
			posY = info.getPosY();
		}

		public Message() {}

		public UUID getUuid() {
			return uuid;
		}

		public ITextComponent getName() {
			return name;
		}

		public Action getAction() {
			return action;
		}

		public boolean shouldMirrorText() {
			return mirror;
		}

		public float getPosX() {
			return posX;
		}

		public float getPosY() {
			return posY;
		}

		public float getPrevPosX() {
			return prevPosX;
		}

		public float getPrevPosY() {
			return prevPosY;
		}

		@Override
		public void fromBytes(ByteBuf buf) {
			PacketBuffer packetBuffer = new PacketBuffer(buf);
			action = packetBuffer.readEnumValue(Action.class);
			uuid = packetBuffer.readUuid();

			switch(action) {
				case ADD:
					String json1 = ByteBufUtils.readUTF8String(buf);
					name = ITextComponent.Serializer.jsonToComponent(json1);

					mirror = packetBuffer.readBoolean();

					posX = packetBuffer.readFloat();
					posY = packetBuffer.readFloat();

					prevPosX = packetBuffer.readFloat();
					prevPosY = packetBuffer.readFloat();
					break;
				case SET_MIRROR:
					mirror = packetBuffer.readBoolean();
					break;
				case SET_NAME:
					String json2 = ByteBufUtils.readUTF8String(buf);
					name = ITextComponent.Serializer.jsonToComponent(json2);
					break;
				case SET_POS:
				case SET_POS_ABSOLUTE:
					posX = packetBuffer.readFloat();
					posY = packetBuffer.readFloat();
					break;
				case REMOVE:
				default:
					break;
			}
		}

		@Override
		public void toBytes(ByteBuf buf) {
			PacketBuffer packetBuffer = new PacketBuffer(buf);
			packetBuffer.writeEnumValue(action);
			packetBuffer.writeUuid(uuid);

			switch(action) {
				case ADD:
					String json1 = ITextComponent.Serializer.componentToJson(name);
					packetBuffer.writeString(json1);

					packetBuffer.writeBoolean(mirror);

					packetBuffer.writeFloat(posX);
					packetBuffer.writeFloat(posY);

					packetBuffer.writeFloat(posX);
					packetBuffer.writeFloat(posY);
					break;
				case SET_MIRROR:
					packetBuffer.writeBoolean(mirror);
					break;
				case SET_NAME:
					String json2 = ITextComponent.Serializer.componentToJson(name);
					packetBuffer.writeString(json2);
					break;
				case SET_POS:
					packetBuffer.writeFloat(posX);
					packetBuffer.writeFloat(posY);
					break;
				case SET_POS_ABSOLUTE:
					packetBuffer.writeFloat(posX);
					packetBuffer.writeFloat(posY);

					packetBuffer.writeFloat(prevPosX);
					packetBuffer.writeFloat(prevPosY);
					break;
				case REMOVE:
				default:
					break;
			}
		}
	}

	public static class Handler implements IMessageHandler<SpellcardInfoPacket.Message, IMessage> {

		@Override
		public IMessage onMessage(Message message, MessageContext ctx) {
			Minecraft.getMinecraft().addScheduledTask(() -> DanmakuCore.proxy.handleSpellcardInfo(message));
			return null;
		}
	}
}

File: src/main/scala/net/katsstuff/danmakucore/network/SpellcardInfoPacket.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network

import java.util.UUID

import net.katsstuff.danmakucore.DanmakuCore
import net.katsstuff.mirror.network.scalachannel.{ClientMessageHandler, Discriminator, MessageConverter}
import net.minecraft.client.network.NetHandlerPlayClient
import net.minecraft.util.text.ITextComponent

sealed trait SpellcardInfoPacket {
  def uuid: UUID
}
case class AddSpellcardInfo(uuid: UUID, name: ITextComponent, mirror: Boolean, posX: Float, posY: Float) extends SpellcardInfoPacket
object AddSpellcardInfo {
  implicit val disc: Discriminator[AddSpellcardInfo] = Discriminator(0.toByte)
}

case class SetSpellcardInfoName(uuid: UUID, name: ITextComponent) extends SpellcardInfoPacket
object SetSpellcardInfoName {
  implicit val disc: Discriminator[SetSpellcardInfoName] = Discriminator(1.toByte)
}

case class SetSpellcardInfoMirror(uuid: UUID, mirror: Boolean) extends SpellcardInfoPacket
object SetSpellcardInfoMirror {
  implicit val disc: Discriminator[SetSpellcardInfoMirror] = Discriminator(2.toByte)
}

case class SetSpellcardInfoPos(uuid: UUID, posX: Float, posY: Float, absolute: Boolean) extends SpellcardInfoPacket
object SetSpellcardInfoPos {
  implicit val disc: Discriminator[SetSpellcardInfoPos] = Discriminator(3.toByte)
}

case class RemoveSpellcardInfo(uuid: UUID) extends SpellcardInfoPacket
object RemoveSpellcardInfo {
  implicit val disc: Discriminator[RemoveSpellcardInfo] = Discriminator(4.toByte)
}
object SpellcardInfoPacket {

  implicit val converter: MessageConverter[SpellcardInfoPacket] = MessageConverter.mkDeriver[SpellcardInfoPacket].apply

  implicit val handler: ClientMessageHandler[SpellcardInfoPacket, Unit] =
    new ClientMessageHandler[SpellcardInfoPacket, Unit] {
      override def handle(netHandler: NetHandlerPlayClient, a: SpellcardInfoPacket): Option[Unit] = {
        scheduler.addScheduledTask(() => DanmakuCore.proxy.handleSpellcardInfo(a))
        None
      }
    }
}
